core/src/main/java/io/aiven/kafka/tieredstorage/RemoteStorageManager.java

```diff
@@ -223,23 +223,26 @@ public Optional<CustomMetadata> copyLogSegmentData(final RemoteLogSegmentMetadat
         final long startedMs = time.milliseconds();
 
         try {
-            TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(
-                Files.newInputStream(logSegmentData.logSegment()), chunkSize);
             SegmentEncryptionMetadataV1 encryptionMetadata = null;
+            final TransformFinisher transformFinisher;
             final boolean requiresCompression = requiresCompression(logSegmentData);
-            if (requiresCompression) {
-                transformEnum = new CompressionChunkEnumeration(transformEnum);
-            }
-            if (encryptionEnabled) {
-                final DataKeyAndAAD dataKeyAndAAD = aesEncryptionProvider.createDataKeyAndAAD();
-                transformEnum = new EncryptionChunkEnumeration(
-                    transformEnum,
-                    () -> aesEncryptionProvider.encryptionCipher(dataKeyAndAAD));
-                encryptionMetadata = new SegmentEncryptionMetadataV1(dataKeyAndAAD.dataKey, dataKeyAndAAD.aad);
+
+            try (final InputStream logSegmentInputStream = Files.newInputStream(logSegmentData.logSegment())) {
+                TransformChunkEnumeration transformEnum = new BaseTransformChunkEnumeration(
+                    logSegmentInputStream, chunkSize);
+                if (requiresCompression) {
+                    transformEnum = new CompressionChunkEnumeration(transformEnum);
+                }
+                if (encryptionEnabled) {
+                    final DataKeyAndAAD dataKeyAndAAD = aesEncryptionProvider.createDataKeyAndAAD();
+                    transformEnum = new EncryptionChunkEnumeration(
+                        transformEnum,
+                        () -> aesEncryptionProvider.encryptionCipher(dataKeyAndAAD));
+                    encryptionMetadata = new SegmentEncryptionMetadataV1(dataKeyAndAAD.dataKey, dataKeyAndAAD.aad);
+                }
+                transformFinisher = new TransformFinisher(transformEnum, remoteLogSegmentMetadata.segmentSizeInBytes());
+                uploadSegmentLog(remoteLogSegmentMetadata, transformFinisher, customMetadataBuilder);
             }
-            final TransformFinisher transformFinisher =
-                new TransformFinisher(transformEnum, remoteLogSegmentMetadata.segmentSizeInBytes());
-            uploadSegmentLog(remoteLogSegmentMetadata, transformFinisher, customMetadataBuilder);
 
             final ChunkIndex chunkIndex = transformFinisher.chunkIndex();
```

Comment:
Should not it be so that the lifecycle of transformFinisher should be same as logSementInputStream? Saving chunkIndex into filed in rather a side effect, it is safe to get it here but the main purpose of the finisher is based on the underlying stream that is already closed afaiu. So I'd rather would not expose finisher from the try related to underlying stream or maybe would even try to make it auto closeable(not sure if it's possible) and close simultaneously with the underlying stream.

Reply:
In general LGTM but not sure if we need a nested try, imo it could be just single one.

Reply:
And please squash

Reply:
Done